File: problem-solving/lc-680-easy-valid-palindrome-ii/1-solution.py
"""
https://leetcode.com/problems/valid-palindrome-ii/

Complexity
----------
Time: O(n)
Space: O(1)
"""

def isPalindrome(s, l, r):
    while l < r:
        if s[l] != s[r]:
            return False
        l += 1
        r -= 1
    return True

class Solution:
    def validPalindrome(self, s: str) -> bool:
        left = 0
        right = len(s) - 1
        while left < right:
            if s[left] != s[right]:
                return isPalindrome(s, left + 1, right) or isPalindrome(s, left, right - 1)
            left += 1
            right -= 1
        return True

File: problem-solving/_lc-0647-medium-palindromic-substrings/1-solution.go
/*
https://leetcode.com/problems/palindromic-substrings/

Complexity
----------
Time: O(n^2)
Space: O(1)
*/

func countPalindrome(s string, left, right int) int {
    count := 0
    for left >= 0 && right < len(s) && s[left] == s[right] {
        left -= 1
        right += 1
        count += 1
    }
    return count
}

func countSubstrings(s string) int {
    count := len(s)
    for mid := 0; mid < len(s); mid++ {
        count += countPalindrome(s, mid - 1, mid + 1)
        count += countPalindrome(s, mid, mid + 1)
    }
    return count
}
